cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -0,0 +1,293 @@
+
+  ASSERT_OK(UnregisterExtensionType(ext_type_->extension_name()));
+  ASSERT_OK(RegisterExtensionType(ext_type_));
+  auto ext_metadata =
+      key_value_metadata({{"ARROW:extension:name", exact_ext_type->extension_name()},
+                          {"ARROW:extension:metadata", serialized_}});
+  auto ext_field = field("f0", exact_ext_type, true, ext_metadata);
+  auto batch = RecordBatch::Make(schema({ext_field}), ext_arr->length(), {ext_arr});
+  std::shared_ptr<RecordBatch> read_batch;
+  RoundtripBatch(batch, &read_batch);
+  CompareBatch(*batch, *read_batch, /*compare_metadata=*/true);
+}
+
+TEST_F(TestExtensionType, RoudtripBatchFromTensor) {
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type_);
+  ASSERT_OK_AND_ASSIGN(auto tensor, Tensor::Make(value_type_, Buffer::Wrap(values_),
+                                                 shape_, {}, dim_names_));
+  EXPECT_OK_AND_ASSIGN(auto ext_arr, exact_ext_type->MakeArray(tensor));
+  ext_arr->data()->type = exact_ext_type;
+
+  ASSERT_OK(UnregisterExtensionType(ext_type_->extension_name()));
+  ASSERT_OK(RegisterExtensionType(ext_type_));
+  auto ext_metadata =
+      key_value_metadata({{"ARROW:extension:name", ext_type_->extension_name()},
+                          {"ARROW:extension:metadata", serialized_}});
+  auto ext_field = field("f0", ext_type_, true, ext_metadata);
+  auto batch = RecordBatch::Make(schema({ext_field}), ext_arr->length(), {ext_arr});
+  std::shared_ptr<RecordBatch> read_batch;
+  RoundtripBatch(batch, &read_batch);
+  CompareBatch(*batch, *read_batch, /*compare_metadata=*/true);
+}
+
+TEST_F(TestExtensionType, ComputeStrides) {
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type_);
+
+  auto ext_type_1 = fixed_shape_tensor(int64(), cell_shape_, {}, dim_names_);
+  auto ext_type_2 = fixed_shape_tensor(int64(), cell_shape_, {}, dim_names_);
+  auto ext_type_3 = fixed_shape_tensor(int32(), cell_shape_, {}, dim_names_);
+  ASSERT_TRUE(ext_type_1->Equals(*ext_type_2));
+  ASSERT_FALSE(ext_type_1->Equals(*ext_type_3));
+
+  auto ext_type_4 = fixed_shape_tensor(int64(), {3, 4, 7}, {}, {"x", "y", "z"});
+  ASSERT_EQ(ext_type_4->strides(), (std::vector<int64_t>{224, 56, 8}));
+  ext_type_4 = fixed_shape_tensor(int64(), {3, 4, 7}, {0, 1, 2}, {"x", "y", "z"});
+  ASSERT_EQ(ext_type_4->strides(), (std::vector<int64_t>{224, 56, 8}));
+
+  auto ext_type_5 = fixed_shape_tensor(int64(), {3, 4, 7}, {1, 0, 2});
+  ASSERT_EQ(ext_type_5->strides(), (std::vector<int64_t>{56, 224, 8}));
+  ASSERT_EQ(ext_type_5->Serialize(), R"({"shape":[3,4,7],"permutation":[1,0,2]})");
+
+  auto ext_type_6 = fixed_shape_tensor(int64(), {3, 4, 7}, {1, 2, 0}, {});
+  ASSERT_EQ(ext_type_6->strides(), (std::vector<int64_t>{56, 8, 224}));
+  ASSERT_EQ(ext_type_6->Serialize(), R"({"shape":[3,4,7],"permutation":[1,2,0]})");
+
+  auto ext_type_7 = fixed_shape_tensor(int64(), {3, 4, 7}, {2, 0, 1}, {});
```

Comment:
```suggestion
  auto ext_type_7 = fixed_shape_tensor(int32(), {3, 4, 7}, {2, 0, 1}, {});
```

Maybe use a different bitwidth in one of them, since the strides are documented to be in bytes and not elements

Reply:
We didn't really discuss the strides in elements vs strides in bytes yet. I've reused logic from Tensor here which defaults to bytes, but I'm not attached to it.

Reply:
To be clear, I am not advocating to change bytes vs elements, just suggesting we have test coverage with different byte sizes, _given_ that we are using bytes.

Reply:
I assumed so. Was just curious if you thought elements makes sense. Let's keep this consistent with `Tensor.strides` then.